kernel/src/schema.rs

```diff
@@ -187,32 +191,27 @@ impl StructField {
             .collect()
     }
 
-    pub fn make_physical(&self, mapping_mode: ColumnMappingMode) -> DeltaResult<Self> {
-        use ColumnMappingMode::*;
-        match mapping_mode {
-            Id => return Err(Error::generic("Column ID mapping mode not supported")),
-            None => return Ok(self.clone()),
-            Name => {} // fall out
-        }
-
-        struct ApplyNameMapping;
-        impl SchemaTransform for ApplyNameMapping {
+    /// Applies physical name mappings to this field
+    ///
+    /// NOTE: Caller affirms that the schema was already validated by
+    /// [`crate::table_features::column_mapping::validate_column_mapping_schema`], to ensure that
+    /// annotations are always and only present when column mapping mode is enabled.
+    pub fn make_physical(&self) -> Self {
+        struct MakePhysical;
+        impl SchemaTransform for MakePhysical {
             fn transform_struct_field<'a>(
                 &mut self,
-                field: Cow<'a, StructField>,
+                field: &'a StructField,
             ) -> Option<Cow<'a, StructField>> {
                 let field = self.recurse_into_struct_field(field)?;
-                match field.get_config_value(&ColumnMetadataKey::ColumnMappingPhysicalName) {
-                    Some(MetadataValue::String(physical_name)) => {
-                        Some(Cow::Owned(field.with_name(physical_name)))
-                    }
-                    _ => Some(field),
-                }
+                Some(Cow::Owned(field.with_name(field.physical_name())))
             }
         }
-
-        let field = ApplyNameMapping.transform_struct_field(Cow::Borrowed(self));
-        Ok(field.unwrap().into_owned())
+        // NOTE: unwrap is safe because the transformer is incapable of returning None
```

Comment:
This is true afaict, but it requires quite a lot of layers to never return `None` since many of the called functions do something like `transform(..)?`. There are many places we could make changes where this assumption would be invalidated.

Would it make sense to return a `Result` or `Option` here and propagate ?

Reply:
I don't love making infallible operations seem fallible... it's one more thing to test that can't actually be tested (because it never happens).

The default implementation of the transform never returns None, and never will (it's a no-op). So we only have to worry about the one method we actually provide here while implementing the trait. And it doesn't directly return `None`. At worst they use `?` on a recursive operation (that never returns `None`).

I don't know a way to generically express that idea in the trait, unfortunately... ideas welcome